compute_tools/src/lsn_lease.rs

```diff
@@ -0,0 +1,187 @@
+use std::{str::FromStr, sync::Arc, thread, time::Duration};
+use utils::id::TenantId;
+use utils::id::TimelineId;
+
+use compute_api::spec::ComputeMode;
+use tracing::{info, warn};
+use utils::{
+    lsn::Lsn,
+    shard::{ShardCount, ShardNumber, TenantShardId},
+};
+
+use crate::compute::{ComputeNode, ComputeState};
+
+/// Spawns a background thread to periodically renew LSN leases for static compute.
+/// Do nothing if the compute is not in static mode.
+pub fn launch_lsn_lease_bg_task_for_static(compute: &Arc<ComputeNode>) {
+    let lsn = {
+        let state = compute.state.lock().unwrap();
+        let spec = state.pspec.as_ref().expect("spec must be set");
+        match spec.spec.mode {
+            ComputeMode::Static(lsn) => lsn,
+            _ => return,
+        }
+    };
+    let compute = compute.clone();
+    thread::spawn(move || {
+        if let Err(e) = lsn_lease_bg_task(compute, lsn) {
+            // TODO: might need stronger error feedback than logging an warning.
+            warn!("lsn_lease_bg_task failed: {e}");
+        }
+    });
+}
+
+fn postgres_configs_from_state(compute_state: &ComputeState) -> Vec<postgres::Config> {
+    let spec = compute_state.pspec.as_ref().expect("spec must be set");
+    let conn_strings = spec.pageserver_connstr.split(',');
+
+    conn_strings
+        .map(|connstr| {
+            let mut config = postgres::Config::from_str(connstr).expect("invalid connstr");
+            if let Some(storage_auth_token) = &spec.storage_auth_token {
+                info!("Got storage auth token from spec file");
+                config.password(storage_auth_token.clone());
+            } else {
+                info!("Storage auth token not set");
+            }
+            config
+        })
+        .collect::<Vec<_>>()
+}
+
+/// Renews lsn lease periodically so static compute are not affected by GC.
+fn lsn_lease_bg_task(compute: Arc<ComputeNode>, lsn: Lsn) -> Result<()> {
+    loop {
+        let (tenant_id, timeline_id, configs) = {
+            let state = compute.state.lock().unwrap();
+
+            let spec = state.pspec.as_ref().expect("spec must be set");
+
+            let configs = postgres_configs_from_state(&state);
+            (spec.tenant_id, spec.timeline_id, configs)
+        };
+
+        let valid_until = acquire_lsn_lease_with_retry(tenant_id, timeline_id, lsn, &configs)
+            .with_context(|| {
+                format!("failed to get lsn lease for {tenant_id}/{timeline_id}@{lsn}")
+            })?;
+
+        let valid_duration = valid_until
+            .duration_since(SystemTime::now())
+            .unwrap_or(Duration::ZERO);
+
+        // Sleep for 60 seconds less than the valid duration but no more than half of the valid duration.
+        let sleep_duration = valid_duration
+            .saturating_sub(Duration::from_secs(60))
+            .max(valid_duration / 2);
+
+        info!(
+            "lsn_lease_request succeeded, sleeping for {} seconds",
```

Comment:
NIT: `lsn_lease_request` here and in the message above looks like a func/method name, but it isn't

Reply:
For this, it might be good to create a `tracing::info_span!` for the thread (near `std::thread::spawn`) and remove all of these prefixes.

Reply:
94a9ee1897be60d20ad1c973e23e893f852d99c5